include/sound/sof/ipc4/header.h

```diff
@@ -0,0 +1,328 @@
+
+#define SOF_IPC4_GL_PIPE_STATE_ID_SHIFT		16
+#define SOF_IPC4_GL_PIPE_STATE_ID_MASK		GENMASK(23, 16)
+#define SOF_IPC4_GL_PIPE_STATE_ID(x)		((x) << SOF_IPC4_GL_PIPE_STATE_ID_SHIFT)
+
+#define SOF_IPC4_GL_PIPE_STATE_SHIFT		0
+#define SOF_IPC4_GL_PIPE_STATE_MASK		GENMASK(15, 0)
+#define SOF_IPC4_GL_PIPE_STATE(x)		((x) << SOF_IPC4_GL_PIPE_STATE_SHIFT)
+
+enum sof_ipc4_sampling_frequency {
+	SOF_IPC4_FS_8000HZ	= 8000,
+	SOF_IPC4_FS_11025HZ	= 11025,
+	SOF_IPC4_FS_12000HZ	= 12000,
+	SOF_IPC4_FS_16000HZ	= 16000,
+	SOF_IPC4_FS_18900HZ	= 18900,
+	SOF_IPC4_FS_22050HZ	= 22050,
+	SOF_IPC4_FS_24000HZ	= 24000,
+	SOF_IPC4_FS_32000HZ	= 32000,
+	SOF_IPC4_FS_37800HZ	= 37800,
+	SOF_IPC4_FS_44100HZ	= 44100,
+	SOF_IPC4_FS_48000HZ	= 48000, /* Default. */
+	SOF_IPC4_FS_64000HZ	= 64000,
+	SOF_IPC4_FS_88200HZ	= 88200,
+	SOF_IPC4_FS_96000HZ	= 96000,
+	SOF_IPC4_FS_176400HZ	= 176400,
+	SOF_IPC4_FS_192000HZ	= 192000,
+};
+
+enum sof_ipc4_bit_depth {
+	SOF_IPC4_DEPTH_8BIT	= 8, /* 8 bits depth */
+	SOF_IPC4_DEPTH_16BIT	= 16, /* 16 bits depth */
+	SOF_IPC4_DEPTH_24BIT	= 24, /* 24 bits depth - Default */
+	SOF_IPC4_DEPTH_32BIT	= 32, /* 32 bits depth */
+	SOF_IPC4_DEPTH_64BIT	= 64, /* 64 bits depth */
+};
+
+enum sof_ipc4_channel_config {
+	/* one channel only. */
+	SOF_IPC4_CHANNEL_CONFIG_MONO = 0,
+	/* L & R. */
+	SOF_IPC4_CHANNEL_CONFIG_STEREO,
+	/* L, R & LFE; PCM only. */
+	SOF_IPC4_CHANNEL_CONFIG_2_POINT_1,
+	/* L, C & R; MP3 & AAC only. */
+	SOF_IPC4_CHANNEL_CONFIG_3_POINT_0,
+	/* L, C, R & LFE; PCM only. */
+	SOF_IPC4_CHANNEL_CONFIG_3_POINT_1,
+	/* L, R, Ls & Rs; PCM only. */
+	SOF_IPC4_CHANNEL_CONFIG_QUATRO,
+	/* L, C, R & Cs; MP3 & AAC only. */
+	SOF_IPC4_CHANNEL_CONFIG_4_POINT_0,
+	/* L, C, R, Ls & Rs. */
+	SOF_IPC4_CHANNEL_CONFIG_5_POINT_0,
+	/* L, C, R, Ls, Rs & LFE. */
+	SOF_IPC4_CHANNEL_CONFIG_5_POINT_1,
+	/* one channel replicated in two. */
+	SOF_IPC4_CHANNEL_CONFIG_DUAL_MONO,
+	/* Stereo (L,R) in 4 slots, 1st stream: [ L, R, -, - ] */
+	SOF_IPC4_CHANNEL_CONFIG_I2S_DUAL_STEREO_0,
+	/* Stereo (L,R) in 4 slots, 2nd stream: [ -, -, L, R ] */
+	SOF_IPC4_CHANNEL_CONFIG_I2S_DUAL_STEREO_1,
+	/* L, C, R, Ls, Rs & LFE., LS, RS */
+	SOF_IPC4_CHANNEL_CONFIG_7_POINT_1,
+};
+
+enum sof_ipc4_interleaved_style {
+	SOF_IPC4_CHANNELS_INTERLEAVED = 0,
+	SOF_IPC4_CHANNELS_NONINTERLEAVED,
+};
+
+enum sof_ipc4_sample_type {
+	SOF_IPC4_MSB_INTEGER = 0, /* integer with Most Significant Byte first */
+	SOF_IPC4_LSB_INTEGER, /* integer with Least Significant Byte first */
+	SOF_IPC4_SIGNED_INTEGER, /* signed integer */
+	SOF_IPC4_UNSIGNED_INTEGER, /* unsigned integer */
+	SOF_IPC4_FLOAT, /* unsigned integer */
+};
+
+struct sof_ipc4_audio_format {
+	enum sof_ipc4_sampling_frequency sampling_frequency;
+	enum sof_ipc4_bit_depth bit_depth;
+	uint32_t ch_map;
+	enum sof_ipc4_channel_config ch_cfg;
+	uint32_t interleaving_style;
+	uint8_t channels_count;
+	uint8_t valid_bit_depth;
+	uint8_t s_type; /* sof_ipc4_sample_type */
+	uint8_t reserved;
+} __attribute__((packed, aligned(1)));
```

Comment:
aligned 1 is different with aligned 4 in fw.    the sizeof struct a { char b;  struct sof_ipc4_audio_format c} will be different between aligned1 and aligned4 since alignment will affect start address.

The best way is defining u32 data and use macro definition to access each field.

Reply:
Thanks @RanderWang I missed that point. Yes, we should have the same aligned as firmware side. Fixed it now.